### PR TITLE
feat: added notifications guide

### PR DESCRIPTION
**Pull request by simonhyll**
Still need to work a bit on it but I'd like some feedback on structure and general direction.

### CODE REVIEW

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/recipe/notifications.mdx]:
```suggestion
In this recipe we'll take a look at sending a native notification to your user. We'll do this by using [`tauri-plugin-notification`](https://github.com/tauri-apps/plugins-workspace/tree/v2/plugins/notification).
```

Where should we link to? I know there's also the read-only mirror repos but I think it's better to reference this one? Might chat back and forth on Discord on this and see what we come up with.

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/recipe/notifications.mdx]: Curious: What does the `tauri plugin add` CLI do? Because if it doesn't do all of this for you it should. If it does I'd have an "automatic" tab and then a "Manual" tab on this page to make sure we cover both strategies.

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/recipe/notifications.mdx]: Alright, another one for the discussion points, but how the heck do we not put ourselves in a position of needed to write every single framework under the sun? I want like one single thing we every "cover". I have a feeling this could be a looooong discussion but think we should have that ASAP. Maybe another on for Discord.

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/recipe/notifications.mdx]: I think I like the phrasing of "frontend" and "backend" and that distinction. Could you do a quick write-up (maybe add it to the CONTRIBUTING guide in this PR) that explains what we mean by that to make it easy to any contributors?

**Review by lorenzolewis — COMMENTED** [review]
Good start and a lot of points already that should be very good discussion points to set us up for a good foundation.

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: Well for now your suggestion of using v2 is the only one that really works. I'm not sure if it's a very permanent link though, we'll have to ask maybeee @lucasfernog for input on this, if the link is permanent enough for use to use and if the read-only mirror is more appropriate

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: It 1. adds cargo dependencies, 2. adds npm dependencies, 3. prints to the terminal what code you need to add to lib.rs or main.rs in order to initialize it

So you could run `tauri plugin add`, but only for specific official plugins, and seeing as it doesn't take care of the Rust part I'd say it's pretty much non-optional to have the `lib.rs` tab visible both in automatic mode and manual mode

I'll see what it looks like using nested tabs. I do think we need both automatic and manual at least.

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/recipe/notifications.mdx]: I sort of hate that they're all spread out across a bunch of repos, but I know there was a long discussion around this and sure it's like that for a good reason, so don't want to beat a dead horse. 

The inconsistency between the core repo's v2 branch and the plugin workspace is something that does drive me a bit crazy (v1, v2, dev, latest, next, 1.x, release, etc.), especially when we're having to hard-code URLs in the docs that could/will change.

So if that's something the dev team would be open to doing now to make it consistent I'd really appreciate it before we write a ton of docs.

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: We can talk on Discord aobut it certainly, but to put my idea somewhere:

The Vanilla one is the cover all version that we always have. Adding more frameworks is optional. I don't even think we should say "these are the X frameworks we support and have available everywhere", if there's an example, there's an example, if there's not, there's not. We get rid of the concept of a list of frameworks we support, we only support vanilla js and have examples for that *always*, but other frameworks might be in there if someone has contributed one. This also plays into the whole community contributing part because you're never going to get rid of people wanting their framework to be represented on the site, so I say let them, we only officially add vanilla js, all others are up to us if we feel like making, and will merge PR's for if the community adds.

That's how I'd do it anyway probably. I understand the tab list can get really long though ofc, but the recipes are explicitly intended to be copy-paste friendly (which is another thing I really want us to add, a copy button to codeblocks).

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: Thanks :) I'll try to write something about it

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/recipe/notifications.mdx]: I think we should look at automating those changes to `lib.rs` or `main.rs` with this command (as well as the plugin migration commands). Anything we can do to automate these things is less of a problem area for people to have issues with (and ask support for).

Plus, for people who aren't comfortable with Rust we really should make that "just work" for them to improve the DX and let them be content in happy JS land.

Let me know if I should open an issue on this upstream.

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/recipe/notifications.mdx]: I agree with this. I think _where_ they put those contributions we can have a think through. I want to make a "Community Resources" component that people can PR links out to their blogs/sites/GH repos/etc. to expand on that topic.

What we actually allow to be contributed to our docs we may need to be more mindful of because a) we're sort of expected to be "experts" if it's in the docs and b) it's not sustainable to cover every single framework.

An option would be to have a page on "common patterns" (reactivity for example) and we could cover that on a framework-by-framework basis and only touch on the Tauri-specific things. Very similar to the idea I think @FabianLars suggested on an FAQ of sorts for frameworks, rather than an all-encompassing guide.

**Comment by lorenzolewis** [comment]
Also, just want to gentle put down a hard no on nested Tabs from a UX perspective 🧡  But more than happy to talk through alternatives if you hit a scenario where you think they'd be useful and see what we can come up with together.

So feel free to commit away with them for this but I wouldn't spend too much time on polishing them up

File: src/content/docs/2/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/recipe/notifications.mdx]: I can also see this would be a very good place to write a concept of "Frontend vs. Backend" and link to it. Maybe we could add a `TODO:` here and slot that into here in the future?

**Comment by lorenzolewis** [comment]
Here's a component we can start re-using and would be good in a guide like this <https://github.com/tauri-apps/tauri-docs/pull/1380>

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: My perception of time is warped, can't believe Github is telling me it has been 5 days ._.

I think this is indeed an upstream issue, we should still advertise it, but it does have some room for improvement

File: src/content/docs/2/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/recipe/notifications.mdx]: Sounds good to me :)

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx:1]: Since this is about base Tauri functionality/plugins I think this belongs directly in guides and not recipes, right? (A splash screen would be a recipe since it builds on top of already existing Tauri functionality but isn't a function/plugin itself)

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]: Should I put together a component like we mentioned in the Docs call that looks like this?



Can put that together in a separate PR pretty quickly if you tell me what info would be good there. I can imagine this:

- Version
- GitHub Repo (plugins monorepo, v2 branch)
- npm page (optional)
- crates.io (optional)
- changelog (I would assume pointing to GitHub somewhere)

Anything else I'm missing?

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]: Would this page contain a demo? I think having those demos as high-up on the page as possible (above the fold) would be best, but would like your feedback.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]:
```suggestion
First install the notification plugin in the backend.
```

"Your backend" sounds funny 😆

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]:
```suggestion
    <TabItem label="Automatic">
    
    Use the [add command in the Tauri CLI](/2/reference/cli/#plugin-add) to add the plugin:
    
    ```bash
    tauri plugin add notification
    ```
    </TabItem>
```

Comment (lorenzolewis) [reply src/content/docs/2/guide/recipe/notifications.mdx]: I think we should list out all the npm/pnpm/cargo/etc. commands. Check out this component to see if it helps simplify that: https://github.com/tauri-apps/tauri-docs/blob/next/src/components/CommandTabs.astro

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]: UX feedback: If they are sequential steps and not mutually exclusive then don't use tabs. If it's "action A AND action B must be done" then list those out without any tabs. If it's "action B OR Action B" then use tabs (like you've done with Automatic vs. Manual here)

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]: This might be good information, but I don't feel it's relevant for the reader here. They don't really care about how it works under the hood. Any of this information I'd capture directly in the source (if it isn't already very clear) where it could be much more relevant for someone there.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/recipe/notifications.mdx]: I really love this type of section. I think this is a really good place we could mark a `TODO:` to write a concept on sanitisation of user input and then link to it from here.

**Review by lorenzolewis — COMMENTED** [review]
Really loving the way it's shaping up. Aside from the individual comments I also thought of this at a higher level:

I found it straightforward the way @lucasfernog framed frontend/backend in the migration guide: https://beta.tauri.app/2/guide/upgrade-migrate/from-tauri-1/#migrate-to-app-plugin

Screenshot 2023-08-16 at 10 11 42

Putting myself in the shoes of a user I think I'd be able to understand if the "JavaScript" or "Rust" section is relevant to me, rather than first thinking of "what is a frontend" vs. "what is a backend" and going from there. Plus the line above that directly says:

> Use in JavaScript **or** Rust project

That makes it very clear to me that I could use either option and guides me through the right logic path for my scenario (also relevant UX feedback on tabs I put in a detailed review comment here).

Welcome any thoughts and feedback on this if you feel differently 🤗

**Comment by lorenzolewis** [comment]
Hope you don't mind, but I put a few general updates on this one to adjust the flow and structure.

I noticed the JS notification API with this one blew up like crazy since v1 which I don't think is the right direction for us. I think we should only have the basics like this guide covers so I'll take that one back to the team

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Yea I feel like as a project with a strong sense for security this sort of section would be really appreciated by people.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: For this specific one there isn't much to say about it, but there are other recipes where it could be more relevant with some form of deep dive. Though then again, the recipes should be short sweet and to the point, so maybe you're right and such deep dives are better done as blog posts.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Good point, I'll get rid of the nasty nested tabs. When I made it I was mostly thinking of tabs like viewing them in VSCode for example, but yea, it's not an IDE, it's a guide, so sequential steps make more sense to have one after the other.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Fixed

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Such a dirty mind! 😂

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Yep this page will have a demo, it used to have one actually but you wanted me to split that component up into its own PR and it hasn't been merged yet 😅 But yes I imagine it being as far up as possible as well, either at the very top or below the description just above the Setup title

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx]: I can't spontaneously think of anything else that would be needed, I think it looks good. 🙂

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/recipe/notifications.mdx:1]: You're the one that wanted me to make it about the notifications 😂 But yes as per our definition of what a recipe is this wouldn't fall under it because it doesn't combine multiple features to create something new, it's really just showing off a single feature, so that feature should stand on its own 2 legs to provide documentation for it.

We could expand the definition however to include anything that's copy-paste friendly and following this specific format? Because plugin docs will almost certainly not be structured quite like this and with this level of copy-paste friendliness, so it's easy for people to jump into a recipe and copy-paste code to get up and running quickly, while reading the actual plugin documentation gets a bit more feature rich and in-depth.

Recipes: short, sweet and to the point, gets you up and running with what's essentially a "Hello World" version of something
Guide/Plugin docs: longer, saltier and with a weird aftertaste, gets you the details of how it works and explores features beyond just the Hello World level

Comment (lorenzolewis) [reply src/content/docs/2/guide/recipe/notifications.mdx:1]: Haha sorry, by "base functionality" I mean what you get with plugins out-of-the-box, not how you compose them together (like a splash screen).

> We could expand the definition however to include anything that's copy-paste friendly and following this specific format? Because plugin docs will almost certainly not be structured quite like this and with this level of copy-paste friendliness, so it's easy for people to jump into a recipe and copy-paste code to get up and running quickly, while reading the actual plugin documentation gets a bit more feature rich and in-depth.

Do you mean that you'd prefer if this is a guide and not a recipe that a guide style would be a bit more detailed and verbose (the weird aftertaste like you say? 😆)

Do you have a few examples of some more details and exploration that could be nice here? Looking at the API references I could see an expansion on these topics:

- Actions: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#action
- Attachments: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#attachment
- Channels: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#channel
- Options: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#options
- Might be some that I'm missing

If that's what you have in mind, then maybe a document structure like this could work:

```md
## Overview (maybe we lump the security heading in here)
## Setup
## Usage
### Send a Notification
### Actions
### Attachments
### Channels
### Options
```

Then within each of the headings under `Usage` you have that copy-pasteable content like you've mentioned? I know we still need to wrap up the frontend/backend vs. js/rust convo which may give a bit more context to this.

The only thing I think to keep in mind if that's the approach you're thinking is that we don't want to turn the guide into a complete reference doc, right? So we might need to play around a tad bit with what we cover vs. what we link over to a reference for.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Added in https://github.com/tauri-apps/tauri-docs/pull/1430, feel free to merge and add it here. Can always do any tweaks on the design later down the road after we see it in context.

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/recipe/notifications.mdx]: Gotcha covered now: https://github.com/tauri-apps/tauri-docs/blob/next/src/components/Demo.astro

File: src/content/docs/2/guide/recipe/notifications.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/recipe/notifications.mdx]: added it in! 🥳

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/notification.mdx]: I wonder if we should just keep these as to-dos for now to really focus on writing the docs?

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/notification.mdx]: Was just reading back through some comments on this PR, `plugin add` doesn't do this but I feel it should (really thinking of those users who don't even touch Rust, this will be a large deterrent for them).

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/notification.mdx:72]: I fully realise this goes against my own UX principals so feel free to rake me across the coals if you don't think it works here @simonhyll 😅

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [inline src/content/docs/2/guide/notification.mdx:140]: @simonhyll what do you think about this sort of structure? I think this could go a really long way to give link-able places for support

Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx:140]: I would put `Security Considerations` further down, feels like the sort of more in-depth topic you read about later on in a post

Not sure what Actions, Attachments or Channels are for? Is it to explain notification concepts or are they general sections that all recipes could have? Options I assume is just a simplified version of what you'd find in the actual API. As long as they aren't too wordy and are copy-paste friendly I'm probably fine with them.

I'm still not overly fond of the terminology of Javascript vs Rust since we should really also have the Rust version avilable using something like Jonas' tauri-sys crate so that the Rust frameworks also get included, because with just JS and Rust (backend) we're kinda leaving them hanging, and I don't see a nice way of adding them without ending up with really annoying names. I'd prefer it if we write about Frontend and Backend as core concepts of Tauri and if people skipped that part then it's not really our problem they refused to read what is almost literally called "the only stuff that's non-optional for you to skip reading". I could also accept Javascript, Rust and Backend I guess because then all frontend languages get included plus one for the backend, but even that isn't completely future proof since we have plans for language bindings at some point.

In the end I defer to your wisdom on the docs in general, so feel free to overrule me on basically anything 😅

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx:72]: I should screenshot this and post on the Astro server 😂 

It certainly makes more sense to do it like that if you structure things without the separated Frontend and Backend sections. As I see it it's not *actually* nested tabs, it's just the ability to switch which package manager you wanna use

Speaking of choosing your package manager, what if when the user switches tabs we save that to LocalStorage so all CommandTabs default to the last one they picked?

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx]: One option is to make the command update lib.rs for them. However, editing source code like that is no easy task, if we just do string replacing then the odds something goes wrong at some point is pretty high, and then we have users with instead of a tiny bit of Rust experience coming to use literally screaming because the black box broke and they don't know how to fix it.

One concept I've toyed with in my mind in the past is creating a tiny Typescript package that allows users to configure their backend with plugins and what not and then we generate the Rust code for them, preferably in something like `.tauri` so it's available, but not so available they'll touch it. The `.gitignore` it, update a `tauri.conf.ts` file, run Node.js on it and badda bing badda boom it generates a the source code of the project. That setup afaik wouldn't even need anything weird like a CLI update to be supported, you could literally just run `pnpm tauri dev` and it's pick up on the project in `.tauri`

Personally I feel source code editing using `plugin add` is very, very risky, and I think the kind of users that can't edit `lib.rs` are the users that need something more reliable than string replacing, so imo my Typescript config idea sounds pretty cool if I do say so myself.

Thoughts?

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx]: Small argument against that would be that if we don't add the demo component now we'll have to rethink the structure later on, and that can be pretty annoying. So I would add the demo component to all of them, and just in the demo component set the text to something like "We're working on a Tauri app, stay tuned!". Something like that. I dunno. I just don't like having to rethink the structure later on.

Bigger argument for it, we do really need to work on content rather than features, I want to start linking people to the beta site as soon as possible.

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx:72]: > I should screenshot this and post on the Astro server 😂

Please do, I'm sure Chris would get a kick out of it 😆 

> Speaking of choosing your package manager, what if when the user switches tabs we save that to LocalStorage so all CommandTabs default to the last one they picked?

I believe that's what they've discussed in https://github.com/withastro/starlight/issues/462. If you get bored and want to dive into Starlight land I'm sure HiDeoo would love to have a contribution based on his issue.

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx:140]: Note that I read your comments from the bottom to the top and just saw your comment earlier explaining what you meant with Actions, Attachments etc. I'm tired, brain no work 😅

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx:140]: Open to whatever you feel would be best for security considerations. I figured if there was anything in-depth that would link out either externally or to a concept and maybe not include it in the guide (unless it really is only very specific to that guide/feature). 

Actions, Attachment, Channels, and Options are part of the plugins APIs, so really just more copy-pasteable examples:

- Actions: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#action
- Attachments: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#attachment
- Channels: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#channel
- Options: https://deploy-preview-1379--tauri-docs-starlight.netlify.app/2/reference/js/notification/#options

RE: Frontend/backend/rust/js how about we put some data-driven decisions in practice and make a poll we can post to Twitter, Discord, and Mastodon and see what comes out? 

> I'm still not overly fond of the terminology of Javascript vs Rust since we should really also have the Rust version avilable using something like Jonas' tauri-sys crate so that the Rust frameworks also get included, because with just JS and Rust (backend) we're kinda leaving them hanging, and I don't see a nice way of adding them without ending up with really annoying names.

Literally have never heard of that until today (looks cool!): https://github.com/JonasKruckenberg/tauri-sys. I do wonder if this is really a more advanced-level usage that's outside of the target audience (i.e. beginners) for docs?

---

A piece that maybe just fell in place for me: do you not like "Rust" because you can have both a Rust frontend (yew, sycamore, etc.) and a Rust backend (application logic, what you're truly referring to as backend) in the same app? Because if that's the case then it just clicked for me and your points are a lot more clear to me now.

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx]: I also think it makes sense to hold off any even attempting to integrate Tauri until a tad bit closer to beta release. I'm personally not all that concerned with the structure point because we're being fairly diligent with how we're structuring the headings so far. 

Another thought I had is that maybe this demo component should really just be a single button and not a whole large box (such as for this guide it would just be a "send notification" button). I don't think we need to parameterise every option because it really is just meant to be a demo. The user gets to have fun in their own app playing with all the levers maybe? 🤔 

But overall I think it really makes sense to hold off a little later down the road when we have a better idea of how the content starts to take shape

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx]: I buy the complicated part, you have to dig into Rust AST which I've never even attempted. But I don't think it's risky per-se. I've seen the DX in other projects (looking at you `astro add`) where the CLI will prompt you with "here's the diff for the changes I'm going to make, does this look right and do you want to continue?".

My 2 arguments to advocate this overall are:
1. JS peeps already get overwhelmed when having to jump into Rust land in 1.0 (myself included sometimes)
2. With everything being moved into plugins for 2.0 we're going to have soooooo many more people touching these things where before it wasn't nearly as common/popular I feel

Tbh what you proposed in TypeScript sounds more complicated with a lot more hoops to jump through the logic path on rather than the CLI doing:

1. Find rust project (I think it already can do this)
2. Find the `lib.rs` file (this is either a known, static file path or can be inferred from `Cargo.toml` I assume?)
3. Find the `run()` fund inside of that (is it always called `run`? If so, then awesome!)
4. A bit of logic magic to find where to insert the plugin initialisation 

I'm sure this exists out there in a lot more places but https://docs.astro.build/en/guides/integrations-guide/svelte/#astro-add-command is what I've seen and used.

File: src/content/docs/2/guide/notification.mdx
Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx:140]: > Note that I read your comments from the bottom to the top and just saw your comment earlier explaining what you meant with Actions, Attachments etc. I'm tired, brain no work 😅

I mean we've only had like a million comments on this one poor PR the past few days 😅

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx]: I can put together a little package for it at some point next week probably, what I have in my head isn't as complicated as it might seem.
1. Yep
2. Static path, but not all that important
3. No need to find anything, my idea overwrites the entire source code in the `.tauri` directory. It's an approach for those that **really** don't wanna touch Rust, so they just configure it in Typescript and each time they build it the source code gets overwritten to reflect what's in the config.
4. Same answer as 3, no magic, just overwriting

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx]: I'm open to whatever as long as we get that DX improvement (one-liner like `tauri plugin add notification`) and @lucasfernog and @wusyong give the thumbs up as I imagine that would be integrated into the core CLI.

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx]: And just to be clear: not a blocking issue to get this merged 😅

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx]: Put in an upstream issue for this one: https://github.com/tauri-apps/tauri/issues/7696

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx]: We can make it just a button, I cant spontaneously think of anything that would *need* configuring, but it's a pretty nice littoe extra thing to have, like adding your own name to the notification showed or what not. Imagine how most UI frameworks show off components letting you tweak various options in the preview, thats kinda what I'm imagining

But yea we'll hold off on it for now and focus on content 👍🏻

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx:140]: Yea exactly, we want to support our buddies over at the Rust based frontend frameworks, so saying "rust" just for the backend isnt as inclusive

And tauri-sys I would say is the more beginner friendly option than the alternative if you develop using e.g. Yew Leptos or Sycamore. You dont even want to know what the alternative looks like, Jonas really did make aomething awesome and at some point we might wanna make it the actual official approach and "take over" his project. It's kind alike the official JS API but for Rust

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx:140]: Could you give some examples of how it's different to use the Rust API in Rust frontend code vs. Rust backend code? Might be better to do it outside of the context of notifications where it's a more practical example, but I think that context is what I'm missing.

Keep in mind that for now we're only covering how to do things with the first-party APIs (I don't think you were thinking of putting `tauri-sys` code in here, and let's talk about it if so, but just wanted to make sure). 

---

I do wonder if "covering all the ways to consume the API" is a similar rabbit hole we got into with "cover all the different JS frontend frameworks"? And if we're wanting to say "here's how you do it in x, then here's a common guide on using Tauri within the context of that". But I doubt we want to NEED to document all of these combinations in every single feature/guide 🤔

**Comment by lorenzolewis** [comment]
This is in a shape where I'm happy to merge it as-is. I know we still have a convo around frontend/backend vs. javascript/rust but this is only a label on a tab here that we can very easily change once we have something with more context to make a better decision on this.

**Review by lorenzolewis — APPROVED** [review]

File: src/content/docs/2/guide/notification.mdx
Comment (simonhyll) [reply src/content/docs/2/guide/notification.mdx:140]: Even if we dont do tauri-sys specifically skipping Rust entirely will leave our beloved rustaceans behind. Using tauri-sys isnt quite the same thing as the framework rabbit hole imo, but if we skip it we should at least offer a "vanilla Rust" example same as we do with vanilla JS

Comment (lorenzolewis) [reply src/content/docs/2/guide/notification.mdx:140]: I agree, but we need the dev team to either do that upstream if it doesn't already, or they need to provide the docs and examples to help us write the guides for it. Right now I didn't find any examples for the Rust side of things to be able to fill it out.

**Pull request merged**
